keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2150 @@
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
+            uses delayed allocation, and
+            was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
+  }
+}
+```
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
+  * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+  * **resource driver** finishes allocation, sets `Allocation` -> claim ready for use by any pod
+* if *pod is pending*:
+  * **scheduler** filters nodes
+  * if *delayed allocation and resource not allocated yet*:
+    * if *at least one node fits pod*:
+      * **scheduler** picks one node, sets `SuitableNodes`, `SelectedNode` and `SelectedUser`
+      * if *resource is available for `SelectedNode`*:
+        * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+        * **resource driver** finishes allocation, sets `Allocation` and the
+          intended user in `ReservedFor` -> claim ready for use and reserved
+          for the pod
+      * else *scheduler needs to know that it must avoid this and possibly other nodes*:
+        * **resource driver** sets `UnsuitableNodes` -> next attempt by scheduler has more information and is more likely to succeed
+    * else *pod cannot be scheduled*:
+      * **scheduler** may trigger freeing some claim (see [pseudo-code above](#coordinating-resource-allocation-through-the-scheduler)) or wait
```

Comment:
what is freeing? which field gets cleaned up?

Reply:
"trigger freeing by setting ResourceClaimStatus.Deallocate to true".

This doesn't occur as often as "allocated" and "(un)used", so I added it to this line.